src/Authoring/WinRT.SourceGenerator/WindowsUIXamlProjectionsAnalyzer.cs

```diff
@@ -0,0 +1,116 @@
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.EnableConcurrentExecution();
+
+        context.RegisterCompilationAction(static context =>
+        {
+            // Are we referencing the Windows.UI.Xaml projections?
+            bool isWindowsUIXamlProjectionsAssemblyReferenced = IsAssemblyReferenced(
+                context.Compilation,
+                "Microsoft.Windows.UI.Xaml",
+                context.CancellationToken);
+
+            // Are we referencing WindowsAppSDK?
+            bool isWindowsAppSDKReferenced = IsAssemblyReferenced(
+                context.Compilation,
+                "Microsoft.Windows.ApplicationModel.WindowsAppRuntime.Projection",
+                context.CancellationToken);
+
+            // Check the MSBuild properties
+            bool useUwp = context.Options.AnalyzerConfigOptionsProvider.GetUseUwp();
+            bool useWinUI = context.Options.AnalyzerConfigOptionsProvider.GetUseWinUI();
+
+            // Error #1: missing 'UseUwp' when using Windows.UI.Xaml projections
+            if (!useUwp && isWindowsUIXamlProjectionsAssemblyReferenced)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(WinRTRules.MissingUseUwpWhenUsingWindowsUIXamlProjections, location: null));
+            }
+
+            // Error #2: using 'UseUwp' when referencing WindowsAppSDK
+            if (useUwp && isWindowsAppSDKReferenced)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(WinRTRules.UsingUseUwpWhenUsingWinAppSDK, location: null));
+            }
+
+            // Error #3: using 'UseWinUI' when referencing Windows.UI.Xaml projections
+            if (useUwp && isWindowsAppSDKReferenced)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(WinRTRules.UsingUseWinUIWhenUsingWindowsUIXamlProjections, location: null));
+            }
+
+            // Error #4: using 'UseUwp' and 'UseWinUI' together
+            if (useUwp && isWindowsAppSDKReferenced)
+            {
+                context.ReportDiagnostic(Diagnostic.Create(WinRTRules.UsingUseUwpAndUseWinUITogether, location: null));
+            }
+        });
+    }
+
+    /// <summary>
+    /// Checks whether a given assembly is referenced.
+    /// </summary>
+    /// <param name="compilation">The input <see cref="Compilation"/> to check.</param>
+    /// <param name="assemblyName">The name of the assembly to check.</param>
+    /// <param name="token">The cancellation token for the operation.</param>
+    /// <returns>Whether <paramref name="compilation"/> references an assembly with a given name.</returns>
+    private static bool IsAssemblyReferenced(
+        Compilation compilation,
+        string assemblyName,
+        CancellationToken token)
+    {
+        foreach (MetadataReference metadataReference in compilation.References)
```

Comment:
@333fred just double checking, we can rely on this always returning all references including from transitive dependencies, right? Basically a scenario we want to check here is, you're referencing library A, and library A is referencing the System XAML projections, or WinAppSDK. That assembly would show up as a reference here even if your project isn't directly referencing them (eg. it's not adding an explicit NuGet package dependency to the package containing that assembly), right?

Reply:
The assemblies being passed to roslyn are resolved by msbuild targets, roslyn doesn't have any knowledge of "transitive dependencies".

Reply:
Right, I know Roslyn doesn't really process dependencies in terms of NuGet packages, but I meant like, practically speaking does this mean I'd always get all assemblies in all those scenarios being passed as references there?

Reply:
Yes, you would get all references passed to the compiler. That _does not_ mean that all references were resolved correctly by msbuild, or that there isn't some other build authoring error that would cause an assembly to be missing.